sns/ut/parity_math_mt_ub.c

```diff
@@ -44,55 +51,62 @@ struct tb_cfg {
 	uint32_t  tc_block_size;
 
 	uint8_t **tc_data;
+	uint8_t **tc_backup;
 	uint8_t **tc_parity;
 	uint8_t  *tc_fail;
 };
 
 static int ub_init(const char *opts M0_UNUSED)
 {
-	srand(1285360231);
 	return 0;
 }
 
 void tb_cfg_init(struct tb_cfg *cfg, uint32_t data_count, uint32_t parity_count,
 		 uint32_t block_size)
 {
 	uint32_t i;
-	uint32_t j;
+	uint32_t fail_count;
 
 	cfg->tc_data_count	= data_count;
 	cfg->tc_parity_count	= parity_count;
 	cfg->tc_fail_count	= data_count + parity_count;
 	cfg->tc_block_size      = block_size;
 
 
-	/* allocate and prepare data */
-	cfg->tc_data = m0_alloc(data_count * sizeof(uint8_t*));
+	/* allocate data */
+	M0_ALLOC_ARR(cfg->tc_data, data_count);
 	M0_ASSERT(cfg->tc_data != NULL);
 
 	for (i = 0; i < data_count; ++i) {
-		cfg->tc_data[i] = m0_alloc(block_size * sizeof(uint8_t));
+		M0_ALLOC_ARR(cfg->tc_data[i], block_size);
 		M0_ASSERT(cfg->tc_data[i] != NULL);
-
-		for (j = 0; j < block_size; ++j)
-			cfg->tc_data[i][j] = (uint8_t)rand();
 	}
 
 	/* allocate parity */
-	cfg->tc_parity = m0_alloc(parity_count * sizeof(uint8_t*));
+	M0_ALLOC_ARR(cfg->tc_parity, parity_count);
 	M0_ASSERT(cfg->tc_parity != NULL);
 
+	/* allocate memory for backup */
+	M0_ALLOC_ARR(cfg->tc_backup, parity_count);
+	M0_ASSERT(cfg->tc_backup != NULL);
+
 	for (i = 0; i < parity_count; ++i) {
-		cfg->tc_parity[i] = m0_alloc(block_size * sizeof(uint8_t));
-		M0_ASSERT(cfg->tc_data[i] != NULL);
+		M0_ALLOC_ARR(cfg->tc_parity[i], block_size);
+		M0_ASSERT(cfg->tc_parity[i] != NULL);
+
+		M0_ALLOC_ARR(cfg->tc_backup[i], block_size);
+		M0_ASSERT(cfg->tc_backup[i] != NULL);
 	}
 
 	/* allocate and set fail info */
-	cfg->tc_fail = m0_alloc(cfg->tc_fail_count * sizeof(uint8_t));
+	M0_ALLOC_ARR(cfg->tc_fail, cfg->tc_fail_count);
 	M0_ASSERT(cfg->tc_fail != NULL);
 
-	for (i = 0; i < parity_count; ++i)
-		cfg->tc_fail[i] = 1; /* maximal possible fails */
+	fail_count = (m0_rnd64(&seed) % parity_count) + 1;
+	for (i = 0; i < fail_count; ++i){
+		uint32_t idx = m0_rnd64(&seed) % cfg->tc_fail_count;
```

Comment:
The result "idx" might be duplicated with previous iteration.
Is that OK?

Reply:
Here, we are generating "fail_count" of failure idx.
But these number might be duplicated. So, the cfg->tc_fail[] array may have fewer number of failure index.

Reply:
Updated the logic to avoid duplicate failure index.